From gcc-patches-return-258497-listarch-gcc-patches=gcc dot gnu dot org at gcc dot gnu dot org Tue Mar 16 13:32:55 2010
Return-Path: <gcc-patches-return-258497-listarch-gcc-patches=gcc dot gnu dot org at gcc dot gnu dot org>
Delivered-To: listarch-gcc-patches at gcc dot gnu dot org
Received: (qmail 25550 invoked by alias); 16 Mar 2010 13:32:54 -0000
Received: (qmail 25534 invoked by uid 22791); 16 Mar 2010 13:32:53 -0000
X-SWARE-Spam-Status: No, hits=-1.1 required=5.0 	tests=AWL,BAYES_00,SUBJ_ALL_CAPS
X-Spam-Check-By: sourceware.org
Received: from mail-bw0-f210.google.com (HELO mail-bw0-f210.google.com) (209.85.218.210)     by sourceware.org (qpsmtpd/0.43rc1) with ESMTP; Tue, 16 Mar 2010 13:32:48 +0000
Received: by bwz2 with SMTP id 2so4002140bwz.30         for <multiple recipients>; Tue, 16 Mar 2010 06:32:45 -0700 (PDT)
Received: by 10.204.14.84 with SMTP id f20mr2900449bka.209.1268746365293;         Tue, 16 Mar 2010 06:32:45 -0700 (PDT)
Received: from s42.loc (85-127-85-230.dynamic.xdsl-line.inode.at [85.127.85.230])         by mx.google.com with ESMTPS id 14sm3626793bwz.14.2010.03.16.06.32.43         (version=TLSv1/SSLv3 cipher=RC4-MD5);         Tue, 16 Mar 2010 06:32:44 -0700 (PDT)
Received: from cow by s42.loc with local (Exim 4.71) 	(envelope-from <rep.dot.nop@gmail.com>) 	id 1NrX2W-0008Nj-5o; Tue, 16 Mar 2010 14:42:32 +0100
From: Bernhard Reutner-Fischer <rep dot dot dot nop at gmail dot com>
To: gcc-patches at gcc dot gnu dot org
Cc: rguenth at gcc dot gnu dot org, 	Bernhard Reutner-Fischer <rep dot dot dot nop at gmail dot com>
Subject: [PATCH] PR32219
Date: Tue, 16 Mar 2010 14:42:27 +0100
Message-Id: <1268746947-32108-1-git-send-email-rep.dot.nop@gmail.com>
Mailing-List: contact gcc-patches-help at gcc dot gnu dot org; run by ezmlm
Precedence: bulk
List-Id: <gcc-patches.gcc.gnu.org>
List-Archive: <http://gcc.gnu.org/ml/gcc-patches/>
List-Post: <mailto:gcc-patches at gcc dot gnu dot org>
List-Help: <mailto:gcc-patches-help at gcc dot gnu dot org>
Sender: gcc-patches-owner at gcc dot gnu dot org
Delivered-To: mailing list gcc-patches at gcc dot gnu dot org

Hi,

As suggested by richi.
regtested on i686-linux-gnu with all default languages and no regressions.
Ok for trunk?

gcc/ChangeLog
2010-03-15  Bernhard Reutner-Fischer  <aldot@gcc.gnu.org>

	PR target/32219
	* varasm.c (default_binds_local_p_1): Weak data is not local.

gcc/testsuite/ChangeLog
2010-03-15  Bernhard Reutner-Fischer  <aldot@gcc.gnu.org>

	PR target/32219
	* gcc.dg/visibility-21.c: New test.

Signed-off-by: Bernhard Reutner-Fischer <rep.dot.nop@gmail.com>
---
 gcc/testsuite/gcc.dg/visibility-21.c |   14 ++++++++++++++
 gcc/varasm.c                         |    8 ++++----
 2 files changed, 18 insertions(+), 4 deletions(-)
 create mode 100644 gcc/testsuite/gcc.dg/visibility-21.c

diff --git a/gcc/testsuite/gcc.dg/visibility-21.c b/gcc/testsuite/gcc.dg/visibility-21.c
new file mode 100644
index 0000000..be7deda
--- /dev/null
+++ b/gcc/testsuite/gcc.dg/visibility-21.c
@@ -0,0 +1,14 @@
+/* PR target/32219 */
+/* { dg-do run } */
+/* { dg-require-visibility "" } */
+/* { dg-options "-fPIC" { target fpic } } */
+
+extern void f() __attribute__((weak,visibility("hidden")));
+extern int puts( char const* );
+int main()
+{
+	if (f)
+		f();
+	return 0;
+}
+
diff --git a/gcc/varasm.c b/gcc/varasm.c
index 6b8222f..6b9269a 100644
--- a/gcc/varasm.c
+++ b/gcc/varasm.c
@@ -6613,6 +6613,10 @@ default_binds_local_p_1 (const_tree exp, int shlib)
   /* Static variables are always local.  */
   else if (! TREE_PUBLIC (exp))
     local_p = true;
+  /* Weak data can be overridden by a strong symbol
+     in another module and so are not local.  */
+  else if (DECL_WEAK (exp))
+    local_p = false;
   /* A variable is local if the user has said explicitly that it will
      be.  */
   else if (DECL_VISIBILITY_SPECIFIED (exp)
@@ -6625,10 +6629,6 @@ default_binds_local_p_1 (const_tree exp, int shlib)
      local.  */
   else if (DECL_VISIBILITY (exp) != VISIBILITY_DEFAULT)
     local_p = true;
-  /* Default visibility weak data can be overridden by a strong symbol
-     in another module and so are not local.  */
-  else if (DECL_WEAK (exp))
-    local_p = false;
   /* If PIC, then assume that any global name can be overridden by
      symbols resolved from other modules, unless we are compiling with
      -fwhole-program, which assumes that names are local.  */
-- 
1.7.0

